### PR TITLE
Miker/150 list page handles search params

### PR DESCRIPTION
**Pull request by mikerid94**

### CODE REVIEW

File: src/admin.controller.ts [hunk 1]
```diff
@@ -28,12 +28,19 @@ import { displayName } from './admin.filters'
...
   // @debt architecture "williamd: this could be made configurable on a per-section basis"
-  return {
-    skip: resultsPerPage * (page - 1),
-    take: resultsPerPage,
-  }
+  query.skip(resultsPerPage * (page - 1)).take(resultsPerPage)
+  return query
+}
+
+function buildSearchQueryOptions(query, searchFields: string[], searchParam: string) {
```
Comment (williamdclt) [inline src/admin.controller.ts]:
```suggestion
function buildSearchQueryOptions(query, searchFields: string[], searchString: string) {
```

File: src/admin.controller.ts [hunk 2]
```diff
@@ -28,12 +28,19 @@ import { displayName } from './admin.filters'
...
+  searchFields.forEach(field => {
+    query.orWhere(`"alias"."${field}" LIKE :name`, {
+      name: `%${searchParam}%`,
+    })
+  })
+  return query
```
Comment (williamdclt) [inline src/admin.controller.ts]: This should definitely go in the adminSite

File: src/admin.controller.ts [hunk 3]
```diff
@@ -98,10 +105,17 @@ export class DefaultAdminController {
     @Req() request: Request,
     @Param() params: AdminModelsQuery,
     @Query('page') pageParam: string = '1',
+    @Query('search') searchParam: string,
   ) {
     const { section, repository, metadata, adminEntity } = await this.getAdminModels(params)
     const page = parseInt(pageParam, 10)
-    const [entities, count] = await repository.findAndCount(getPaginationQueryOptions(page))
+    let query = repository.createQueryBuilder('alias')
```
Comment (williamdclt) [inline src/admin.controller.ts]:
```suggestion
    let query = repository.createQueryBuilder('column')
```

File: src/admin.controller.ts [hunk 4]
```diff
@@ -98,10 +105,17 @@ export class DefaultAdminController {
...
+    query = buildPaginationQueryOptions(query, page)
+
+    if (searchParam) {
+      query = buildSearchQueryOptions(query, adminEntity.searchFields, searchParam)
+    }
+    const [entities, count] = await query.getManyAndCount()
```
Comment (williamdclt) [inline src/admin.controller.ts]: I'd like to not have to worry about the query at all in the controller, this should go in the adminSite

File: exampleApp/src/user/user.admin.ts [hunk 5]
```diff
@@ -2,9 +2,11 @@ import { AdminEntity, PasswordWidget } from 'nestjs-admin'
 import { User } from './user.entity'
 
 export class UserAdmin extends AdminEntity {
+  newAction = () => console.log('hello')
```
Comment (williamdclt) [inline exampleApp/src/user/user.admin.ts]: What's that?

File: exampleApp/src/user/user.admin.ts [hunk 6]
```diff
@@ -2,9 +2,11 @@ import { AdminEntity, PasswordWidget } from 'nestjs-admin'
...
   entity = User
   listDisplay = ['id', 'firstName', 'lastName', 'email', 'createdDate']
   searchFields = ['firstName', 'lastName', 'email', 'description']
+  listActions = [{ label: 'NEW BUTTON', action: this.newAction }]
```
Comment (williamdclt) [inline exampleApp/src/user/user.admin.ts]: ??

File: src/adminEntity.ts [hunk 7]
```diff
@@ -91,6 +90,37 @@ abstract class AdminEntity {
     validateFieldsExist(this, 'searchFields', this.metadata)
     validateFieldsAreNotRelation(this, 'searchFields', this.metadata)
   }
+
+  buildSearchQueryOptions(
+    query: SelectQueryBuilder<unknown>, // @debt typing "miker: can I type this better?"
```
Comment (williamdclt) [inline src/adminEntity.ts]: If `unknown` fits the bill, no need for an @debt

File: src/adminSite.ts [hunk 8]
```diff
@@ -74,6 +74,17 @@ class DefaultAdminSite {
...
+  async getChangeList(adminEntity: AdminEntity, page: number, searchParam: string) {
+    const repository = this.getRepository(adminEntity.entity)
+    const alias = 'column'
+
+    let query = repository.createQueryBuilder(alias)
+    query = adminEntity.buildPaginationQueryOptions(query, page)
+    query = adminEntity.buildSearchQueryOptions(query, { alias, searchParam })
+
+    return await query.getManyAndCount()
+  }
```
Comment (williamdclt) [inline src/adminSite.ts]: This should also be in the adminEntity

File: src/adminSite.ts [hunk 9]
```diff
@@ -74,6 +74,17 @@ class DefaultAdminSite {
     }
   }
 
+  async getChangeList(adminEntity: AdminEntity, page: number, searchParam: string) {
+    const repository = this.getRepository(adminEntity.entity)
+    const alias = 'column'
```
Comment (williamdclt) [inline src/adminSite.ts]:
```suggestion
    const alias = 'entity'
```

File: src/adminEntity.ts [hunk 10]
```diff
@@ -91,6 +90,37 @@ abstract class AdminEntity {
...
+    options: { alias: string; searchParam: string },
```
Comment (williamdclt) [inline src/adminEntity.ts]:
```suggestion
    alias: string, searchParam: string
```

File: src/adminSite.ts [hunk 11]
```diff
@@ -74,6 +74,17 @@ class DefaultAdminSite {
     }
   }
 
+  async getChangeList(adminEntity: AdminEntity, page: number, searchParam: string) {
```
Comment (williamdclt) [inline src/adminSite.ts]:
```suggestion
  async getEntityList(adminEntity: AdminEntity, page: number, searchParam: string) {
```

File: src/utils/typeormSwitch.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import * as ProdTORM from 'typeorm' // for prod
+
+let TORM = ProdTORM
+
+if (process.env.NESTJS_ADMIN_TYPEORM_PATH) {
+  TORM = require(process.env.NESTJS_ADMIN_TYPEORM_PATH)
+}
+
+export = TORM
```
Comment (williamdclt) [inline src/utils/typeormSwitch.ts]: `typeormSwitch` isn't a great name. Maybe `typeormProxy`? And we need a big fat comment at the top of this file to explain why we're doing this

File: src/utils/typeormSwitch.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import * as ProdTORM from 'typeorm' // for prod
```
Comment (williamdclt) [inline src/utils/typeormSwitch.ts]:
```suggestion
import * as NestJSTypeORM from 'typeorm'
```

File: tslint.json [hunk 14]
```diff
@@ -18,7 +18,8 @@
     "trailing-comma": false,
     "max-classes-per-file": false,
     "only-arrow-functions": false,
-    "curly": false
+    "curly": false,
+    "no-var-requires": false
```
Comment (williamdclt) [inline tslint.json]: Let's not allow requires everywhere just because we needed a dirty hack once

File: src/adminEntity.ts (hunk 10 above)
Comment (mikerid94) [reply src/adminEntity.ts]: why is this better?

File: src/adminSite.ts (hunk 8 above)
Comment (mikerid94) [reply src/adminSite.ts]: pourquoi?

File: src/adminEntity.ts [hunk 15]
```diff
@@ -91,6 +90,38 @@ abstract class AdminEntity {
...
+  }
+
+  buildPaginationQueryOptions(query: SelectQueryBuilder<unknown>, page: number) {
+    query.skip(this.resultsPerPage * (page - 1)).take(this.resultsPerPage)
+    return query
+  }
+
+  getManyAndCount(query: SelectQueryBuilder<unknown>) {
+    return query.getManyAndCount()
+  }
```
Comment (williamdclt) [inline src/adminEntity.ts]: That doesn't look like it belongs here? What abstraction are you trying to provide here?

File: src/adminSite.ts [hunk 16]
```diff
@@ -74,6 +74,17 @@ class DefaultAdminSite {
...
+  async getEntityList(adminEntity: AdminEntity, page: number, searchParam: string) {
+    const repository = this.getRepository(adminEntity.entity)
+    const alias = 'entity'
+
+    let query = repository.createQueryBuilder(alias)
+    query = adminEntity.buildPaginationQueryOptions(query, page)
+    query = adminEntity.buildSearchQueryOptions(query, { alias, searchParam })
+
+    return await adminEntity.getManyAndCount(query)
+  }
```
Comment (williamdclt) [inline src/adminSite.ts]: This should be in the adminEntity

File: src/utils/typeormProxy.ts [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+//
+// In our dev env there are two typeorm libraries:
+// The first is in the library itself in /node_modiules
+// The second is in exampleApp/node_modules
+//
+// This was causing issues with conditions within 'typeorm' e.g. "instanceof {Class}" comparisons
+// This broke some features of 'typeorm' when we were developing using exampleApp
+// So we have added the NESTJS_ADMIN_TYPEORM_PATH env variable to set which 'typeorm' library to use in 'start:debug' command
+//
+// This should not affect prod
```
Comment (williamdclt) [inline src/utils/typeormProxy.ts]:
```suggestion

```

File: src/adminEntity.ts (hunk 10 above)
Comment (williamdclt) [reply src/adminEntity.ts]: This "options" thing doesn't make semantic sense to me, what does it represent? Does it provide an abstraction, or is it just a bunch of parameters necessary for the function execution?

File: src/adminEntity.ts [hunk 18]
```diff
@@ -91,6 +90,39 @@ abstract class AdminEntity {
     validateFieldsExist(this, 'searchFields', this.metadata)
     validateFieldsAreNotRelation(this, 'searchFields', this.metadata)
   }
+
+  buildSearchQueryOptions(query: SelectQueryBuilder<unknown>, alias: string, searchParam: string) {
```
Comment (williamdclt) [inline src/adminEntity.ts]:
```suggestion
  protected buildSearchQueryOptions(query: SelectQueryBuilder<unknown>, alias: string, searchParam: string) {
```

File: src/adminEntity.ts [hunk 19]
```diff
@@ -91,6 +90,39 @@ abstract class AdminEntity {
...
+            })
+            return qb
+          }),
+        ),
+      )
+    }
+    return query
+  }
+
+  buildPaginationQueryOptions(query: SelectQueryBuilder<unknown>, page: number) {
```
Comment (williamdclt) [inline src/adminEntity.ts]:
```suggestion
  protected buildPaginationQueryOptions(query: SelectQueryBuilder<unknown>, page: number) {
```

File: src/adminEntity.ts [hunk 20]
```diff
@@ -91,6 +90,39 @@ abstract class AdminEntity {
...
+    query.skip(this.resultsPerPage * (page - 1)).take(this.resultsPerPage)
+    return query
+  }
+
+  getManyAndCount(page: number, searchString: string) {
```
Comment (williamdclt) [inline src/adminEntity.ts]:
```suggestion
  getEntityList(page: number, searchString: string): {entities: object[], count: number} {
```

File: src/adminEntity.ts [hunk 21]
```diff
@@ -91,6 +90,39 @@ abstract class AdminEntity {
...
+    const alias = this.name
+    let query = this.repository.createQueryBuilder(alias)
+    query = this.buildPaginationQueryOptions(query, page)
+    query = this.buildSearchQueryOptions(query, alias, searchString)
+    return query.getManyAndCount()
```
Comment (williamdclt) [inline src/adminEntity.ts]:
```suggestion
    const [enttieis, count] = query.getManyAndCount()
    return {entities, count}
```

File: src/adminSite.ts [hunk 22]
```diff
@@ -74,6 +74,10 @@ class DefaultAdminSite {
     }
   }
 
+  async getEntityList(adminEntity: AdminEntity, page: number, searchSting: string) {
```
Comment (williamdclt) [inline src/adminSite.ts]:
```suggestion
  async getEntityList(adminEntity: AdminEntity, page: number, searchString: string) {
```

File: src/tests/changelist.spec.ts [hunk 23]
```diff
@@ -49,6 +53,23 @@ describe('changelist', () => {
...
+    })
+    const userData2 = createTestUser({
+      firstName: 'Jane',
+      lastName: 'Doe',
+    })
+    const userData3 = createTestUser({
+      firstName: 'Will',
+      lastName: 'Duc',
+    })
+    const userRepository: Repository<User> = app.get(getRepositoryToken(User))
```
Comment (williamdclt) [inline src/tests/changelist.spec.ts]: You'l have to use the entitiyManager here

Comment (mikerid94) [reply src/tests/changelist.spec.ts]: yup

**Pull request merged**